server/controllers/admin.py

```diff
@@ -835,6 +836,95 @@ def student_view(cid, email):
                            student=student, enrollment=enrollment,
                            assignments=assignments)
 
+@admin.route("/course/<int:cid>/<string:email>/<int:aid>/timeline")
+@is_staff(course_arg='cid')
+def assignment_timeline(cid, email, aid):
+    courses, current_course = get_courses(cid)
+
+    assign = Assignment.query.filter_by(id=aid, course_id=cid).one_or_none()
+    if not assign or not Assignment.can(assign, current_user, 'grade'):
+        flash('Cannot access assignment', 'error')
+        return abort(404)
+
+    student = User.lookup(email)
+    if not student.is_enrolled(cid):
+        flash("This user is not enrolled", 'warning')
+
+    user_ids = assign.active_user_ids(student.id)
+
+    analytics = (db.session.query(Backup, Message)
+                       .outerjoin(Message)
+                       .filter(Backup.submitter_id.in_(user_ids),
+                               Backup.assignment_id == assign.id,
+                               Message.kind == "analytics")
+                       .order_by(Backup.created.asc())
+                       .all())
+
+    unlock_started_q = {}
+    started_questions = {}
+    solved_questions = {}
+    submitter_counts = {}
+    history, timeline = [], []
+    last_q = (None, False, 0)
```

Comment:
I think this `last_q` makes it harder to understand what exactly is happening when we modify/read it. Being a dictionary or a namedtuple would make the usage clearer. 

`x[2]` vs `x['count']` 

I think namedtuple is the way to go